controllers/group_totals.go

```diff
@@ -56,48 +54,132 @@ func InstaCountHandler(rw http.ResponseWriter, r *http.Request) {
+				equery += "patient.gender=" + gender + "&"
+			}
+
+			//Condition
+			if coding.System == "http://loinc.org" && coding.Code == "11450-4" {
+				cquery += "code=" + characteristic.ValueCodeableConcept.Coding[0].System + "|" + characteristic.ValueCodeableConcept.Coding[0].Code + "&"
+				intersectRequired = true
+			}
+
+			//Encounter
+			if coding.System == "http://loinc.org" && coding.Code == "46240-8" {
+				equery += "type=" + characteristic.ValueCodeableConcept.Coding[0].System + "|" + characteristic.ValueCodeableConcept.Coding[0].Code + "&"
+				intersectRequired = true
+			}
+		}
 	}
-	json.NewEncoder(rw).Encode(qr)
-}
 
-func InstaCountAllHandler(rw http.ResponseWriter, r *http.Request) {
-	decoder := json.NewDecoder(r.Body)
-	group := &fhirmodels.Group{}
-	err := decoder.Decode(group)
-	if err != nil {
-		http.Error(rw, err.Error(), http.StatusInternalServerError)
+	pquery = TrimSuffix(pquery, "&")
+	cquery = TrimSuffix(cquery, "&")
+	equery = TrimSuffix(equery, "&")
+
+	searcher := search.NewMongoSearcher(server.Database)
+
+	var pResultIDs []struct {
+		ID string `bson:"_id"`
+	}
+	pSearchQuery := search.Query{Resource: "Patient", Query: pquery}
+	pQ := searcher.CreateQuery(pSearchQuery)
+	pQ.Select(bson.M{"_id": 1}).All(&pResultIDs)
+	pids := make([]string, len(pResultIDs))
+	for i := range pResultIDs {
+		pids[i] = pResultIDs[i].ID
 	}
 
-	pipelineMap := make(map[string]models.Pipeline)
+	if intersectRequired {
+		type patientContainer struct {
+			ID string `bson:"referenceid"`
+		}
 
-	pipelineMap["patients"] = models.NewPipeline(group)
-	pipelineMap["encounters"] = models.NewEncounterPipeline(group)
-	pipelineMap["conditions"] = models.NewConditionPipeline(group)
+		var cResultIDs []struct {
+			Patient patientContainer `bson:"patient"`
+		}
+		cSearchQuery := search.Query{Resource: "Condition", Query: cquery}
+		cQ := searcher.CreateQuery(cSearchQuery)
+		cQ.Select(bson.M{"patient.referenceid": 1}).All(&cResultIDs)
+		cids := make([]string, len(cResultIDs))
+		for i := range cResultIDs {
+			cids[i] = cResultIDs[i].Patient.ID
+		}
 
-	resultMap := make(map[string]int)
+		var eResultIDs []struct {
+			Patient patientContainer `bson:"patient"`
+		}
+		eSearchQuery := search.Query{Resource: "Encounter", Query: equery}
+		eQ := searcher.CreateQuery(eSearchQuery)
+		eQ.Select(bson.M{"patient.referenceid": 1}).All(&eResultIDs)
+		eids := make([]string, len(eResultIDs))
+		for i := range eResultIDs {
+			eids[i] = eResultIDs[i].Patient.ID
+		}
 
-	for pipelineType, pipeline := range pipelineMap {
-		qr, err := pipeline.ExecuteCount(server.Database)
-		if err != nil && err != mgo.ErrNotFound {
-			http.Error(rw, err.Error(), http.StatusInternalServerError)
-			return
+		firstIntersect := make(map[string]bool)
+		secondIntersect := make(map[string]bool)
+		finalIntersect := make(map[string]bool)
+
+		for _, pid := range pids {
+			firstIntersect[pid] = true
+		}
+
+		for _, cid := range cids {
+			if firstIntersect[cid] == true {
```

Comment:
Actually, now that I think of it, is `firstIntersect` even needed?  If a `condition` criteria or `encounter` criteria is used, then we probably don't need to do the initial patient query, because we only care about patients returned by the `condition` and/or `encounter` query (since they have the patient demographic filter chained onto them already).

In fact, we could probably gain some efficiency by having something like:
```
var pids []string
if hasConditionCriteria && hasEncounterCriteria {
  // pids is intersection of patient ids from cQuery and eQuery
} else if hasConditionCriteria {
  // pids is just patient ids from cQuery
} else if hasEncounterCriteria {
  // pids is just patient ids from eQuery
} else {
  // pids is patient ids from pQuery
}